# Module 2 - NumPy Mathematical Operations

## 1. Sum

### Definition

The Sum function calculates the total of all elements in a NumPy array.

It is useful for finding total marks, total sales, total revenue, and other numerical values.

### Real-world Example

If we have the marks of students, Sum can be used to calculate the total marks.


In [1]:
import numpy as np

scores = np.array([78, 85, 92, 67, 88])

total = np.sum(scores)

print(total)

410


###  Explanation

The `np.sum()` function adds all the values in the array.

The total of the given scores is 410.

### Best Practices

Use `np.sum()` when you need to calculate the total of numerical values in an array.

### Common Mistakes

Forgetting to import NumPy or using the wrong array while calculating the sum.

## 2. Mean

### Definition

The Mean is the average value of all numbers in a NumPy array.

It is calculated by adding all values and dividing the total by the number of values.

### Real-world Example

The Mean can be used to find the average marks of students.

### AI/ML Example

Mean is commonly used to understand the average value of numerical data during data analysis and preprocessing.

In [2]:
average = np.mean(scores)

print(average)

82.0


###  Explanation

The `np.mean()` function calculates the average of all values in the `scores` array.

The average score is 82.0.

### Best Practices

Use `np.mean()` when you need to find the average of numerical values.

### Common Mistakes

Forgetting that the mean can be affected by very high or very low values in the data.

## 3. Median

### Definition

The Median is the middle value after arranging the data in ascending order.

If there are two middle values, their average is taken.

### Real-world Example

Median can be used to find the middle value of student marks.


### AI/ML Example

Median is useful when analyzing data that contains extreme values or outliers.

In [3]:
median = np.median(scores)

print(median)

85.0


### Explanation

The `np.median()` function finds the middle value of the sorted data.

The median score is 83.0.

### Best Practices

Use median when the data may contain extreme values.

### Common Mistakes

Assuming the median is always the same as the mean.

## 4. Variance

### Definition

Variance measures how far the values are spread from the mean.

A small variance means the values are closer together.

A large variance means the values are more spread out.


In [4]:
variance = np.var(scores)

print(variance)

77.2


###  Explanation

The `np.var()` function calculates the variance of the scores.

The variance is 79.484375.

### Best Practices

Use variance to understand how much the data varies from the mean.

### Common Mistakes

Confusing variance with standard deviation.

## 5. Standard Deviation

### Definition

Standard Deviation measures how much the values differ from the mean.

It is the square root of variance.

### Real-world Example

Standard deviation can show how much student marks vary from the average mark.

### Business Example

A company can use standard deviation to understand variations in sales.

### AI/ML Example

Standard deviation is commonly used in data analysis and feature scaling.

In [5]:
std = np.std(scores)

print(std)

8.78635305459552


###  Explanation

The `np.std()` function calculates the standard deviation of the scores.

The standard deviation is approximately 8.92.

### Best Practices

Use standard deviation when you need to understand the spread of data.

### Common Mistakes

Confusing standard deviation with variance.

## 6. Minimum

### Definition

The Minimum function returns the smallest value in an array.

### Real-world Example

We can find the lowest marks scored by a student.

### Business Example

A company can find the lowest sales value from its sales data.

### AI/ML Example

Minimum values can be used during data analysis and preprocessing.

In [6]:
minimum = np.min(scores)

print(minimum)

67


###  Explanation

The `np.min()` function finds the smallest value in the array.

The minimum score is 67.

### Best Practices

Use `np.min()` when you need to find the lowest value.

### Common Mistakes

Confusing the minimum value with its index position.

## 7. Maximum

### Definition

The Maximum function returns the largest value in an array.

### Real-world Example

We can find the highest marks scored by a student.

### Business Example

A company can find the highest sales value from its sales data.

### AI/ML Example

Maximum values are useful for understanding the range of numerical data.

In [7]:
maximum = np.max(scores)

print(maximum)

92


###  Explanation

The `np.max()` function finds the largest value in the array.

The maximum score is 95.

### Best Practices

Use `np.max()` when you need to find the highest value.

### Common Mistakes

Confusing the maximum value with its index position.

## 8. Argmin

### Definition

Argmin returns the index position of the smallest value in an array.

`min()` returns the value, while `argmin()` returns its position.

### Real-world Example

We can find the position of the student with the lowest score.

### Business Example

A company can find the position of the product with the lowest sales.

### AI/ML Example

Argmin can be used when we need to identify the position of the smallest numerical value.

In [8]:
index = np.argmin(scores)

print(index)

3


###  Explanation

The smallest score is 67.

Its index position is 3 because NumPy uses zero-based indexing.

### Best Practices

Use `argmin()` when you need the position of the smallest value.

### Common Mistakes

Confusing the index returned by `argmin()` with the minimum value itself.

## 9. Argmax

### Definition

Argmax returns the index position of the largest value in an array.

`max()` returns the value, while `argmax()` returns its position.

### Real-world Example

We can find the position of the student with the highest score.

### Business Example

A company can identify the product with the highest sales.

### AI/ML Example

Argmax is commonly used to find the position of the highest probability or score in Machine Learning.

In [9]:
index = np.argmax(scores)

print(index)

2


###  Explanation

The largest score is 95.

Its index position is 5.

### Best Practices

Use `argmax()` when you need the position of the largest value.

### Common Mistakes

Confusing the returned index with the maximum value.

## 10. Percentile

### Definition

A percentile indicates the value below which a certain percentage of observations fall.

Common percentiles include the 25th, 50th, and 75th percentiles.

The 50th percentile is the median.

### Real-world Example

Percentiles can be used to understand how student scores are distributed.

### Business Example

Companies can use percentiles to analyze salaries, sales, or customer spending.

### AI/ML Example

Percentiles can be useful for understanding data distribution and detecting unusual values.

In [10]:
print(np.percentile(scores, 25))
print(np.percentile(scores, 50))
print(np.percentile(scores, 75))

78.0
85.0
88.0


###  Explanation

The 25th percentile is 76.75.

The 50th percentile is 83.0, which is the median.

The 75th percentile is 89.0.

### Best Practices

Use percentiles to understand the distribution and position of values in data.

### Common Mistakes

Confusing a percentile with a percentage.

## 11. Dot Product

### Definition

The Dot Product multiplies corresponding elements of two arrays and then adds the results.

It is an important operation in mathematics and Machine Learning.

### Real-world Example

The dot product can be used to combine values such as quantity and price.

### Business Example

A company can use a dot product to calculate total revenue from product quantities and prices.

### AI/ML Example

Dot products are widely used in Machine Learning calculations, including weights and features.

In [11]:
a = np.array([2, 3, 4])
b = np.array([5, 6, 7])

result = np.dot(a, b)

print(result)

56


###  Explanation

The dot product is calculated as:

(2 × 5) + (3 × 6) + (4 × 7) = 56

The `np.dot()` function performs this calculation.

### Best Practices

Make sure the two arrays have compatible sizes before calculating the dot product.

### Common Mistakes

Confusing dot product with normal element-wise multiplication.

## 12. Matrix Multiplication

### Definition

Matrix multiplication combines two matrices to produce a new matrix.

The number of columns in the first matrix must match the number of rows in the second matrix.

### Real-world Example

Matrix multiplication can be used to combine different sets of numerical information.

### Business Example

Businesses can use matrix operations for calculations involving products, customers, and numerical data.

### AI/ML Example

Matrix multiplication is an important operation in Machine Learning and neural networks.

In [12]:
A = np.array([[1, 2],
              [3, 4]])

B = np.array([[5, 6],
              [7, 8]])

result = A @ B

print(result)

[[19 22]
 [43 50]]


###  Explanation

The `@` operator performs matrix multiplication.

The two matrices are multiplied according to matrix multiplication rules.

### Best Practices

Check the dimensions of the matrices before multiplying them.

### Common Mistakes

Confusing matrix multiplication with element-wise multiplication.

## 13. Transpose

### Definition

Transpose changes the rows of a matrix into columns and the columns into rows.

### Real-world Example

A table containing information by rows can be rearranged so that the same information is represented by columns.

### Business Example

Transpose can be used to rearrange numerical business data for analysis.

### AI/ML Example

Transpose is commonly used in matrix calculations and Machine Learning algorithms.

In [13]:
A = np.array([[1, 2, 3],
              [4, 5, 6]])

print(A.T)

[[1 4]
 [2 5]
 [3 6]]


###  Explanation

The `.T` attribute transposes the matrix.

The original 2 × 3 matrix becomes a 3 × 2 matrix.

### Best Practices

Check the original shape before transposing a matrix.

### Common Mistakes

Confusing transpose with reshaping.

## 14. Basic Linear Algebra Operations

### Definition

Linear Algebra deals with vectors, matrices, and mathematical operations on them.

NumPy provides the `linalg` module for common linear algebra operations.

### Real-world Example

Linear algebra can be used to represent and solve relationships between numerical values.

### Business Example

Businesses can use matrix calculations for numerical analysis and optimization problems.

### AI/ML Example

Linear algebra is a fundamental part of Machine Learning, neural networks, dimensionality reduction, and data transformations.

 ### 14.1 Matrix Addition

Two matrices are added element by element.

### Best Practices

Make sure both matrices have the same shape before adding them.

### Common Mistakes

Trying to add matrices with incompatible shapes.

In [16]:
A = np.array([[1, 2],
              [3, 4]])

B = np.array([[5, 6],
              [7, 8]])

print(A + B)

[[ 6  8]
 [10 12]]


### 14.2 Matrix Subtraction

Matrix subtraction subtracts corresponding elements of two matrices.

### Best Practices

Use matrices with the same shape for subtraction.

### Common Mistakes

Confusing matrix subtraction with matrix multiplication.

In [17]:
print(A - B)

[[-4 -4]
 [-4 -4]]


### 14.3 Scalar Multiplication

Scalar multiplication multiplies every element of a matrix by the same number.

### Best Practices

Use scalar multiplication when the same adjustment needs to be applied to all values.

### Common Mistakes

Confusing scalar multiplication with matrix multiplication.

In [18]:
print(A * 3)

[[ 3  6]
 [ 9 12]]


### 14.4 Matrix Multiplication

The `@` operator performs matrix multiplication.

### Best Practices

Check that the matrix dimensions are compatible before multiplication.

### Common Mistakes

Using element-wise multiplication when matrix multiplication is required.

In [19]:
print(A @ B)

[[19 22]
 [43 50]]


### 14.5 Determinant

The determinant is a single value calculated from a square matrix.

It helps determine important properties of a matrix, such as whether it has an inverse.

### Best Practices

Use `np.linalg.det()` only with square matrices.

### Common Mistakes

Trying to calculate the determinant of a non-square matrix.

In [21]:
print(np.linalg.det(A))

-2.0000000000000004


## Advantages

- NumPy provides fast numerical calculations.
- It makes mathematical operations simple.
- It supports arrays and matrices efficiently.
- It provides useful statistical functions.
- It includes many linear algebra operations.

## Limitations

- NumPy mainly focuses on numerical data.
- Large arrays can require significant memory.
- Some mathematical operations require knowledge of array dimensions.
- It is not designed for complex table-based data management.

## Interview Questions and Answers

### 1. What is NumPy?

NumPy is a Python library used for numerical computing and array operations.

### 2. What does np.mean() do?

It calculates the average value of the elements in an array.

### 3. What is the difference between min() and argmin()?

min() returns the smallest value, while argmin() returns its index.

### 4. What is the difference between variance and standard deviation?

Variance measures the spread using squared differences from the mean. Standard deviation is the square root of variance.

### 5. What is a percentile?

A percentile represents the value below which a certain percentage of observations fall.

### 6. What is a dot product?

A dot product multiplies corresponding values of two arrays and adds the results.

### 7. What is matrix multiplication?

Matrix multiplication combines two matrices according to matrix multiplication rules.

### 8. What does transpose do?

Transpose changes rows into columns and columns into rows.

### 9. What is the determinant of a matrix?

The determinant is a single value calculated from a square matrix and provides information about its properties.

### 10. Where is linear algebra used in AI/ML?

Linear algebra is used in Machine Learning, neural networks, matrix calculations, and techniques such as PCA.

## Self Reflection

I learned how NumPy can be used to perform mathematical and statistical operations on numerical data.

I practiced measures such as mean, median, variance, standard deviation, minimum, maximum, and percentiles.

I also learned basic linear algebra operations such as dot product, matrix multiplication, transpose, determinant.